template-coq/src/g_template_coq.ml4

```diff
@@ -47,99 +45,99 @@ let ltac_apply (f : Value.t) (args: Tacinterp.Value.t list) =
 END;;
 
 VERNAC COMMAND EXTEND Make_recursive CLASSIFIED AS SIDEFF
     | [ "Quote" "Recursively" "Definition" ident(name) ":=" constr(def) ] ->
       [ let (evm,env) = Pfedit.get_current_context () in
-	let def, uctx = Constrintern.interp_constr env evm def in
-	let trm = Constr_quoter.TermReify.quote_term_rec env (EConstr.to_constr evm def) in
-	ignore(Declare.declare_definition
-	  ~kind:Decl_kinds.Definition name
-	  (trm, Monomorphic_const_entry (UState.context_set uctx))) ]
+        let (evm, def) = Constrintern.interp_open_constr env evm def in
+        let (evm, pgm) = EConstr.fresh_global env evm Template_monad.ptmQuoteRecDefinition in
+        let pgm = Constr.mkApp (EConstr.to_constr evm pgm, [|Constr_quoter.TemplateCoqQuoter.quote_ident name; Constr.mkRel 0; EConstr.to_constr evm def|]) in
+        run_template_program env evm pgm ]
 END;;
 
 VERNAC COMMAND EXTEND Unquote_vernac CLASSIFIED AS SIDEFF
     | [ "Make" "Definition" ident(name) ":=" constr(def) ] ->
       [ let (evm, env) = Pfedit.get_current_context () in
-	let (trm, uctx) = Constrintern.interp_constr env evm def in
-	let evm, trm = Denote.denote_term evm (EConstr.to_constr evm trm) in
-	let _ = Declare.declare_definition
-                  ~kind:Decl_kinds.Definition
-                  name
-                  (trm, Monomorphic_const_entry (Evd.universe_context_set evm)) in
-        () ]
-END;;
-
-VERNAC COMMAND EXTEND Unquote_vernac_red CLASSIFIED AS SIDEFF
-    | [ "Make" "Definition" ident(name) ":=" "Eval" red_expr(rd) "in" constr(def) ] ->
-      [ let (evm, env) = Pfedit.get_current_context () in
-	let (trm, uctx) = Constrintern.interp_constr env evm def in
-        let evm = Evd.from_ctx uctx in
-        let (evm,rd) = Tacinterp.interp_redexp env evm rd in
-	let (evm,trm) = Quoter.reduce env evm rd (EConstr.to_constr evm trm) in
-        let evm, trm = Denote.denote_term evm trm in
-	let _ = Declare.declare_definition ~kind:Decl_kinds.Definition name
-                  (trm, Monomorphic_const_entry (Evd.universe_context_set evm)) in
-        () ]
+        let (evm, def) = Constrintern.interp_open_constr env evm def in
+        let (evm, pgm) = EConstr.fresh_global env evm Template_monad.ptmMkDefinition in
+        let pgm = Constr.mkApp (EConstr.to_constr evm pgm, [|Constr_quoter.TemplateCoqQuoter.quote_ident name; EConstr.to_constr evm def|]) in
+        run_template_program env evm pgm ]
 END;;
 
 VERNAC COMMAND EXTEND Unquote_inductive CLASSIFIED AS SIDEFF
     | [ "Make" "Inductive" constr(def) ] ->
-      [ let (evm,env) = Pfedit.get_current_context () in
-	let (body,uctx) = Constrintern.interp_constr env evm def in
-        Run_template_monad.declare_inductive env evm (EConstr.to_constr evm body) ]
+      [ let (evm, env) = Pfedit.get_current_context () in
+        let (evm, def) = Constrintern.interp_open_constr env evm def in
+        let (evm, pgm) = EConstr.fresh_global env evm Template_monad.ptmMkInductive in
```

Comment:
Is it feasible to implement this directly rather than building a `constr` that represents the term, and then interpreting it? In particular, could you use the combinators inside of `plugin_core` directly?

Reply:
I don't know. Probably I could use your combinators, but I also like the idea that the command is exactly defined by `tmMkInductive` and that it is completely transparent for the user. What do you think?

Reply:
Sure.